Add functionality for EnvironmentSettings

varzac opened the pull request:
EnvironmentSettings allow for epoxy components to take their value from
operating system environment variables.  The order of preference for the
settings variables is OS environment variable value, yaml configuration
value, default value.

The SettingInstance object was not necessary for the way it was being used.  Since it was only used in an intermediary way within the from_dependencies method, it was removed and the functionality folded into the Settings themselves.  This allows for the proper control of the precedence of values for EnvironmentSettings

varzac committed:
Add functionality for EnvironmentSettings

EnvironmentSettings allow for epoxy components to take their value from
operating system environment variables.  The order of preference for the
settings variables is OS environment variable value, yaml configuration
value, default value.

posborne commented on epoxy/settings.py:
```diff
@@ -117,3 +105,24 @@ def encode(self, value):
         return value
+
+
+class EnvironmentSetting(BaseSetting):
```
Although the others lack it presently, I think we should probably start adding more full docstrings for public items like this that explain usage.  For `EnvironmentSettting`, I think describing the precedence ordering of the following is probably the key piece of information a user would need to understand how this settings works:

* Default specified in code
* Setting value specified in YML configuration
* Environment variable value

varzac committed:
Add documentation. Make the default/set_value precedence of settings explicit

This changes the default value of settings to use a singleton of NO_VALUE
on creation.  It then determines where to return the set value or the
default value depending by checking against the singleton.

posborne commented:
Looks good to me.  I think we'll need a larger effort to improve the documentation, but this is good for now.
